scipy/special/cephes/stirling.c

```diff
@@ -14,30 +14,29 @@
  *  to be partitioned and k is the number of non-empty subsets.
  *  The values for n < 0 or k < 0 are interpreted as 0. If you
  *
- * ACCURACY: The method returns an unsigned integer type and
- *  the size of the type 32 bits.
+ * ACCURACY: Returns correctly rounded result to double precision
+ * if result does not overflow an unsigned 64 bit int. Currently
+ * returns nan if result overflows usigned 64 bit int, with plan
+ * to eventually use an asymptotic approximation in these cases.
  *
  *
  * NOTE: this file is *NOT* part of the ceph distribution and was
  *  added by Lucas Roberts in 2023 after opening this issue
  *  https://github.com/scipy/scipy/issues/17890
  */
 #include "mconf.h"
-#include <limits.h>
+#include <stdint.h>
 #include <stdlib.h>
 
-long stirling2(int n, int k){
-    if (n == 0 && k == 0){
-        return 1L;
-    }
-    if (k <= 0 || k > n || n <= 0){
-        return 0L;
-    }
-    int arraySize = k <= n - k + 1 ? k : n - k + 1;
-    long *curr = malloc(arraySize * sizeof(long));
+static uint64_t stirling2(uint64_t n, uint64_t k){
```

Comment:
I'm not sure we can support these in special at the moment, they would be `long long` correct?

Reply:
The way this is set up, this function is completely internal to this file, note the `static`. It's not exposed in special. `stirling2_approx` uses this function internally but casts the result to double, which is handled. See [here](https://github.com/scipy/scipy/blob/0a017f2f79cfd27f7d7d59a59ed759b00174f737/scipy/special/_hyp2f1.pxd#L225-L233) for an example where I used `uint64_t` in an internal function which is used in the kernel of a ufunc.

Reply:
> The way this is set up, this function is completely internal to this file, note the `static`. It's not exposed in special. `stirling2_approx` uses this function internally but casts the result to double, which is handled. See [here](https://github.com/scipy/scipy/blob/0a017f2f79cfd27f7d7d59a59ed759b00174f737/scipy/special/_hyp2f1.pxd#L225-L233) for an example where I used `uint64_t` in an internal function which is used in the kernel of a ufunc.

ok. I understand now, thanks for clarifying.